edb/edgeql/compiler/stmt.py

```diff
@@ -230,7 +232,146 @@ def compile_GroupQuery(
         context=expr.context)
 
 
+TBase = TypeVar('TBase', bound=qlast.Base)
+
+
+def subject_substitute(ast: TBase, new_subject: qlast.Expr) -> TBase:
+    ast = copy.deepcopy(ast)
```

Comment:
Can this be done by setting up a `__subject__` anchor before compiling the node instead?

Reply:
Yes and no:
 * Yes, but we have to use the subjectexpr twice, on each side of the equation, so we'd have to do it by compiling it twice, once with each side as an anchor, and then putting the results of those in as anchors. I think it would work just fine but it's not really any simpler, I guess. (Well, not any less code. Maybe simpler.)
 * But handling object constraints is going to require a souped-up version of the substitution approach, since we'll need to be able to pull things out of the shape based on path references

Reply:
(So my plan is to stick with the current approach)